WoodFisher Recipe Tutorial (#282)
neuralmagic/sparseml: integrations/pytorch/tutorials/woodfisher_pruning.md

limitations under the License.
-->

# WoodFisher Pruning with SparseML

This tutorial shows how to use the WoodFisher pruning algorithm in a SparseML PyTorch
pruning recipe. The WoodFisher algorithm was introduced in the paper 
[WoodFisher: Efficient Second-Order Approximation for Neural Network Compression](https://arxiv.org/pdf/2004.14340.pdf)
by Sidak Pal Singh and Dan Alistarh.

## Overview
The WoodFisher algorithm introduces an efficient process for approximating accurate
second-order information about a model's weights that can be used to solve for optimal
weights to prune.  The method uses the weights' first-order information
(gradients) to approximate the Inverse-Hessian matrix that is used to score weights
for pruning.

This method was shown to achieve SoTA results for pruning tasks on
image classification models such as ResNet and MobileNet. Detailed information of the
methods and results can be found in the [paper](https://arxiv.org/pdf/2004.14340.pdf).

WoodFisher pruning can be enabled by a single modifier in a SparseML recipe. This tutorial
will demonstrate how to adapt an existing SparseML pruning recipe to do this.

> 

Reviewer: ```suggestion

## Need Help?
For Neural Magic Support, sign up or log in to get help with your questions in our **Tutorials channel:** [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 

```

Author: added to bottom

Reviewer: For the template, we're intentionally upfront letting people know that if they need any help before they actually start the tutorial, there's a method to do it. We then remind them at the end as a wrap-up, to not forget about the offer to help.

Author: moved it up

## WoodFisher Pruning Recipes
To use the WoodFisher algorithm in a SparseML pruning recipe, the `MFACPruningModifier`
must be used in place of the existing pruning modifier(s).  the `MFACPruningModifier`
contains the same parameters as SparseML's magnitude pruning modifiers for controlling
epoch range, sparsity levels, and schedule, so these values may remain
unchaged when adjusting a recipe.

> The `MFACPruningModifier` implements a global pruning scheme however, so all

Reviewer: ```suggestion
The `MFACPruningModifier` implements a global pruning scheme, however, so all
```

parameters to be pruned should be listed under the `params` argument. If the previous
recipe has multiple pruning modifiers, all of their prameters should be merged into a
single `MFACPruningModifier`.

WoodFisher specific hyperparameters can be specified under the `mfac_options` parameter.

> Values should be listed as a YAML dictionary.  Full detials about these hyperparameters

Reviewer: ```suggestion
Values should be listed as a YAML dictionary.  Full details about these hyperparameters
```

are listed in the next section of this tutorial.

### Example Recipe Section

The following is an example `MFACPruningModifier` to be used in place of other

> PruningModifiers in a recipe:

Reviewer: PruningModifiers is one string here right?

Author: changing to two lower case strings for consistency. It may be one in the future

stored when that sparsity level (between 0.0 and 1.0) is reached. If a
dictionary, then 0.0 must be included as a key for the base number of gradients
to store (i.e. {0: 64, 0.5: 128, 0.75: 256}).

Storing gradients can be expensive, as for a dense model, each additional gradient
sample stored requires about the same memory that the entire model needs. This is why
the dictionary option allows for more gradients to be stored as the model gets more
sparse.

If a WoodFisher pruning run is unexpectedly killed, the reason could likely be that
the gradient storage requirements exceeded the system's RAM. A safe rule of thumb for
initial number of gradients is the number should be no greater than 1/4 of the
available CPU RAM divided by the model size.

### fisher_block_size
To limit the computational cost of calculating second order information, the WoodFisher
algorithm may compute a block diagonal matrix of a certain block size that is
sufficient for generating the necessary information for pruning.

The `fisher_block_size` specifies this block size.  If using GPUs to perform the
WoodFisher computations, the GPUs should have `num_grads * fisher_block_size` extra

> memory duing training so each block can be stored and computed sequentially on a GPU.

Reviewer: ```suggestion
memory during training so each block can be stored and computed sequentially on a GPU.
```

The default block size is 2000, and generally block sizes between 1000 and 10000 may be
ideal. If `None` is provided, the full matrix will be computed without blocks.

### available_gpus
`available_gpus` is a list of GPU devices names to perform the WoodFisher computation
with. If not provided, computation will be done on the CPU.

## Wrap-Up
This tutorial shows how to use WoodFisher pruning by swapping in a single modifier
to a SparseML pruning recipe. Everything else for pruning and exporting stays the same.
More papers and results will be posted as they become available.

Open an

> [issue](https://github.com/neuralmagic/sparseml/issues) with any questions or bugs.

Reviewer: ```suggestion

For Neural Magic Support, sign up or log in to get help with your questions in our **Tutorials channel:** [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 

```

> [issue](https://github.com/neuralmagic/sparseml/issues) with any questions or bugs.

Reviewer: ```suggestion
[issue](https://github.com/neuralmagic/sparseml/issues) with any bugs or feature requests.

```suggestion

For Neural Magic Support, sign up or log in to get help with your questions in our **Tutorials channel:** [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 

```
```
